cordova-common/src/FileUpdater.js

```diff
@@ -0,0 +1,389 @@
+
+    var targetFullPath = path.join(rootDir, targetPath);
+    var targetStats = fs.existsSync(targetFullPath) ? fs.statSync(targetFullPath) : null;
+    var sourceStats = null;
+
+    if (sourcePath) {
+        // A non-null source path was specified. It should exist.
+        var sourceFullPath = path.join(rootDir, sourcePath);
+        if (!fs.existsSync(sourceFullPath)) {
+            throw new Error("Source path does not exist: " + sourcePath);
+        }
+
+        sourceStats = fs.statSync(sourceFullPath);
+
+        // Create the target's parent directory if it doesn't exist.
+        var parentDir = path.dirname(targetFullPath);
+        if (!fs.existsSync(parentDir)) {
+            shell.mkdir("-p", parentDir);
+        }
+    }
+
+    return updatePathWithStats(
+        rootDir, targetPath, targetStats, sourcePath, sourceStats, force, log);
+}
+
+/**
+ * Updates files and directories based on a mapping from target paths to source paths. Targets
+ * with null sources in the map are deleted.
+ *
+ * @param {string|null} rootDir Root directory (such as a project) to which target and source
+ *     path parameters are relative, or null if the paths are absolute. The rootDir is omitted
+ *     from any logged paths, to make the logs easier to read.
+ * @param {object} pathMap A dictionary mapping from target paths to source paths.
+ * @param {boolean} force If target and source are both files, and the force flag is not
+ *     set, then the file will not be copied unless the source is newer than the target.
+ * @param {function} [log] Optional logging callback that takes a string message describing any
+ *     file operations that are performed.
+ * @return {boolean} true if any changes were made, or false if the force flag is not set
+ *     and everything was up to date
+ */
+function updatePaths(rootDir, pathMap, force, log) {
+    if (!pathMap || typeof(pathMap) !== "object") {
+        throw new Error("An object mapping from target paths to source paths is required.");
+    }
+
+    var updated = false;
+
+    for (var targetPath in pathMap) {
+        var sourcePath = pathMap[targetPath];
+        updated = updatePath(rootDir, targetPath, sourcePath, force, log) || updated;
+    }
+
+    return updated;
+}
+
+/**
+ * Updates a target directory with merged files and subdirectories from source directories.
+ *
+ * @param {string|null} rootDir Root directory (such as a project) to which target and source
+ *     path parameters are relative, or null if the paths are absolute. The rootDir is omitted
+ *     from any logged paths, to make the logs easier to read.
+ * @param {string} targetDir Destination directory to be updated. If it does not exist, it will be
+ *     created. If it exists, newer files from source directories will be copied over, and files
+ *     missing in the source directories will be deleted.
+ * @param {string|string[]} sourceDirs Source directory or array of source directories to be
```

Comment:
Most uses of JSDoc that I've seen omit those parentheses. And VS Code correctly parses it without parentheses.

---
In reply to: [60399853](https://github.com/apache/cordova-lib/pull/429#discussion_r60399853) [](ancestors = 60399853)